src/libraries/System.ComponentModel.Primitives/src/ILLink/ILLink.LinkAttributes.xml

```diff
@@ -0,0 +1,55 @@
+<linker>
+  <assembly fullname="System.ComponentModel.Primitives">
+    <type fullname="System.ComponentModel.BrowsableAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
+    <type fullname="System.ComponentModel.CategoryAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
+    <type fullname="System.ComponentModel.DescriptionAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
+    <type fullname="System.ComponentModel.DesignerAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
+    <type fullname="System.ComponentModel.DesignerCategoryAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
+    <type fullname="System.ComponentModel.DesignOnlyAttribute">
+      <attribute internal="RemoveAttributeInstances" />
+    </type>
+    <type fullname="System.ComponentModel.DisplayNameAttribute">
```

Comment:
Will this break code like https://github.com/dotnet/aspnetcore/blob/5c897ac0827d6cb306f7864b148d49ccf484a089/src/Mvc/Mvc.DataAnnotations/src/DataAnnotationsMetadataProvider.cs#L89 ?

Reply:
I don't know what the code does as it looks like some sort of fallback. If this code is ever trimmed and needed you should get IL2045 [warning](https://github.com/mono/linker/blob/master/docs/error-codes.md#il2045-trim-analysis-attribute-type-is-being-referenced-in-code-but-the-linker-was-instructed-to-remove-all-instances-of-this-attribute-if-the-attribute-instances-are-necessary-make-sure-to-either-remove-the-linker-attribute-xml-portion-which-removes-the-attribute-instances-or-override-the-removal-by-using-the-linker-xml-descriptor-to-keep-the-attribute-type-which-in-turn-keeps-all-of-its-instances)

Reply:
> If this code is ever trimmed and it needs the attribute IL2045 warning should show up

But what is a user supposed to do in this case? If we unconditionally say "remove DisplayNameAttribute", and there is code using that attribute, they get a warning. But what do they do with it?

There are other attributes in this list that are used at runtime as well:

* ReadOnlyAttribute
* EditorAttribute
* CategoryAttribute
* DescriptionAttribute

Reply:
> But what is a user supposed to do in this case

Follow the warning message and add XML descriptor to keep the attribute type, that will overwrite the removal. 

> There are other attributes in this list that are used at runtime as well:

That's the reason why I made the PR they are used but not relevant for anything else than perhaps some windows designers

Reply:
> but not relevant for anything else than perhaps some windows designers

They are used by public APIs. I don't think we should be forcing apps that use certain APIs to add XML descriptors in order for their app to be trim compatible.

Here's another example in ML.NET that uses the `DescriptionAttribute` - https://github.com/dotnet/machinelearning/blob/c2f8ed5c9c478194fa6c7e387bc9f8fc43e4e0bc/src/Microsoft.ML.Core/Utilities/Utils.cs#L1246-L1265

How much size is this change actually saving? Would it be more appropriate to only do it for WASM?

Reply:
As I wrote in the description we could do this conditionally but are you really expecting someone to run things like VS designers on the self-contained trimmed configuration?

Reply:
@stephentoub and I pointed out 2 different scenarios that are not VS designers.

Reply:
I'm not familiar with the code to see if it's used at design-time or runtime but have no objection to make this conditional for few RIDs only. I'll update the PR.

Reply:
What do you think about removing

* DisplayNameAttribute
* DescriptionAttribute
* ReadOnlyAttribute

from the list as well? We found 2 cases of code in the `dotnet` org using the first 2 attributes. And here is code using `ReadOnlyAttribute` in `System.Data.Common`:

https://github.com/dotnet/runtime/blob/f9bb1673708ca840da0e71f9a9444ea9b0d31911/src/libraries/System.Data.Common/src/System/Data/Common/DbConnectionStringBuilder.cs#L423-L433

(`reflected.IsReadOnly` on line 433 is driven by the ReadOnlyAttribute)

Reply:
> What do you think about removing

They don't seem to be used often so that should be fine but does not give me good feeling about the solution. I was thinking also about enhancing linker to have more than only forced attribute removal but also conditional removal which we could use here but also elsewhere (e.g. for ObsoleteAttribute). It would remove the attribute if no typeref exists to the attribute type. We could then keep the removal without building for different RIDs. What do you think?

/cc @vitek-karas

Reply:
> I was thinking also about enhancing linker to have more than only forced attribute removal but also conditional removal which we could use here but also elsewhere

How similar/different is that to https://github.com/mono/linker/issues/1378?

In general, I think what you are proposing would be good. Here's my summary of it, to make sure I understand the idea:

* `dotnet/runtime` builds a list of attributes that can be removed if no code is directly referencing them
* If the ILLinker sees code referencing one of those attributes (like the `reflected.IsReadOnly` usage above), then it leaves all the instances of that attribute in place and doesn't issue a warning

Reply:
I ... don't like attributes 😉 ... I guess the proposed functionality in the linker is OK - it's kind of similarly bad as the one we have already (the unconditional remove).

@marek-safar I know we're trying to remove everything we don't need - but I think it would be worth getting at least some data for this - how much it helps in a normal Xamarin app - or something like that.

Reply:
> In general, I think what you are proposing would be good. Here's my summary of it, to make sure I understand the idea:

Yep, that's right. I'll turn that into linker proposal.

> how much it helps in a normal Xamarin app 
It seems like the world has moved and a lot of these attributes usage was removed. It's not worth the hustle so I'll close the PR.

@eerhardt @stephentoub I appreciate the feedback, thanks